flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/AbstractTwoInputStreamOperatorWithTTL.scala

```diff
@@ -39,6 +38,15 @@ import org.apache.flink.table.runtime.types.CRow
   * state to cleanup and what further action to take.
   *
   * This class takes care of maintaining at most one timer per key.
+  *
+  * <p><b>IMPORTANT NOTE TO USERS:</b> When extending this class, do not use processing time
+  * timers in your business logic. The reason is that:
+  *
+  * 1) if your timers collide with clean up timers and you delete them, then state
+  * clean-up will not be performed, and
+  *
+  * 2) (this one is the reason why this class does not allow to override the onProcessingTime())
+  * the onProcessingTime with your logic would be also executed on each clean up timer.
   */
 @PublicEvolving
 abstract class AbstractTwoInputStreamOperatorWithTTL(
```

Comment:
Maybe rename it to `TwoInputStreamOperatorWithStateRetention`? I think name TTL is confusing with TTL in state backend while "state retention" is something different.